devito/mpi/distributed.py

```diff
@@ -600,17 +616,53 @@ class CustomTopology(tuple):
     """
 
     def __new__(cls, items, input_comm):
+        # Keep track of nstars and already defined decompositions
         nstars = len([i for i in items if i == '*'])
-        if nstars > 0:
-            if input_comm.size % nstars != 0:
-                raise ValueError("Invalid `topology` for given nprocs")
-            if any(i not in ('*', 1) for i in items):
-                raise ValueError("Custom topology must be only 1 or *")
-
-            v = input_comm.size // nstars
-            processed = [i if i == 1 else v for i in items]
-        else:
+
+        # If no stars exist we are ready
+        if nstars == 0:
             processed = items
+        else:
+            # Init decomposition list
+            processed = [1] * len(items)
+
+            # Get star and integer indices
+            int_pos = [i for i, item in enumerate(items) if isinstance(item, int)]
+            int_vals = [item for item in items if isinstance(item, int)]
+            star_pos = [i for i, item in enumerate(items) if not isinstance(item, int)]
+
+            # Decompose the processes remaining for allocation to prime factors
+            alloc_procs = np.prod([i for i in items if i != '*'])
+            remprocs = int(input_comm.size // alloc_procs)
+            prime_factors = primefactors(remprocs)
+
+            star_i = -1
+            dd_list = [1] * nstars
+
+            # Start by using the max prime factor at the first starred position,
+            # then cyclically-iteratively decompose as evenly as possible until
+            # decomposing to the number of `remprocs`
+            while remprocs != 1:
```

Comment:
nitpicking: if one passes a dummy comm with .size = 0, this remains entrapped in an infinite loop. Safer to use : `while remprocs > 1`

Also, I see an inconsisten use of `_` in variable names

Reply:
Switched to `rem_procs`.

Reply:
I think passing 0 is bad use, maybe we can jump to the assert statement  directly ?
Asking cutom topo on 0 ranks looks bad to me